atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchEventCacheImpl.java

```diff
@@ -0,0 +1,223 @@
+    }
+
+    @VisibleForTesting
+    LockWatchEventCacheImpl(ClientLockWatchEventLog eventLog) {
+        this.eventLog = eventLog;
+        timestampMap = new TimestampToVersionMap();
+    }
+
+    @Override
+    public Optional<IdentifiedVersion> lastKnownVersion() {
+        return eventLog.getLatestKnownVersion();
+    }
+
+    @Override
+    public void processStartTransactionsUpdate(
+            Set<Long> startTimestamps,
+            LockWatchStateUpdate update) {
+        Optional<IdentifiedVersion> latestVersion = processEventLogUpdate(update);
+
+        latestVersion.ifPresent(
+                version -> startTimestamps.forEach(timestamp -> timestampMap.putStartVersion(timestamp, version)));
+
+        getEarliestVersion().ifPresent(eventLog::removeOldEntries);
+    }
+
+    @Override
+    public void processGetCommitTimestampsUpdate(
+            Collection<TransactionUpdate> transactionUpdates,
+            LockWatchStateUpdate update) {
+        Optional<IdentifiedVersion> latestVersion = processEventLogUpdate(update);
+
+        latestVersion.ifPresent(version -> transactionUpdates.forEach(
+                transactionUpdate -> checkConditionOrThrow(!timestampMap.putCommitUpdate(transactionUpdate, version),
+                        "start timestamp missing from map")));
+    }
+
+    @Override
+    public CommitUpdate getCommitUpdate(long startTs) {
+        Optional<IdentifiedVersion> startVersion = timestampMap.getStartVersion(startTs);
+        Optional<CommitInfo> maybeCommitInfo = timestampMap.getCommitInfo(startTs);
+
+        checkConditionOrThrow(!maybeCommitInfo.isPresent() || !startVersion.isPresent(),
+                "start or commit info not processed for start timestamp");
+
+        CommitInfo commitInfo = maybeCommitInfo.get();
+
+        ClientLogEvents update = eventLog.getEventsBetweenVersions(startVersion, commitInfo.commitVersion());
+
+        if (update.clearCache()) {
+            return ImmutableInvalidateAll.builder().build();
+        }
+
+        return constructCommitUpdate(commitInfo, update.events());
+    }
+
+    @Override
+    public TransactionsLockWatchEvents getEventsForTransactions(
+            Set<Long> startTimestamps,
+            Optional<IdentifiedVersion> startVersion) {
+        Preconditions.checkArgument(!startTimestamps.isEmpty(), "Cannot get events for empty set of transactions");
+        Map<Long, IdentifiedVersion> timestampToVersion = getTimestampMappings(startTimestamps);
+        IdentifiedVersion endVersion = Collections.max(timestampToVersion.values(),
+                Comparator.comparingLong(IdentifiedVersion::version));
+        return eventLog.getEventsBetweenVersions(startVersion, endVersion).map(timestampToVersion);
+    }
+
+    @Override
+    public void removeTransactionStateFromCache(long startTimestamp) {
+        timestampMap.remove(startTimestamp);
```

Comment:
Again, this is where you should clear out your stale events.

Reply:
Done